lib/src/sql/latlng.rs

```diff
@@ -0,0 +1,226 @@
+		self.0
+	}
+}
+
+impl fmt::Display for CardinalDegree {
+	fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+		// todo: precision?
+		write!(f, "{:.2}°{}", self.1, self.0)
+	}
+}
+
+/// Represents a latitude and longitude in decimal degrees.
+/// https://en.wikipedia.org/wiki/Decimal_degrees
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(crate) struct DecimalDegrees {
+	vertical: CardinalDegree,
+	horizontal: CardinalDegree,
+}
+
+impl From<(CardinalDegree, CardinalDegree)> for DecimalDegrees {
+	fn from(value: (CardinalDegree, CardinalDegree)) -> Self {
+		DecimalDegrees {
+			vertical: value.0,
+			horizontal: value.1,
+		}
+	}
+}
+
+impl Into<Point> for DecimalDegrees {
+	/// Converts the decimal degrees into a [`Point`] by taking into account
+	/// the direction of each degree.
+	fn into(self) -> Point {
+		Point::new(
+			self.horizontal.value() * self.horizontal.direction(),
+			self.vertical.value() * self.vertical.direction(),
+		)
+	}
+}
+
+impl fmt::Display for DecimalDegrees {
+	fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+		write!(f, "{} {}", self.vertical, self.horizontal)
+	}
+}
+
+/// Parses vertical directions (North or South). Callers may want to
+/// parse vertical separate from horizontal because you can't parse
+/// two vertical directions, i.e. N 0.00° S 0.00° is invalid.
+fn vertical_dir(input: &str) -> IResult<&str, CardinalDirection> {
+	let (i, direction) = alt((
+		map(char('N'), |_| CardinalDirection::North),
+		map(char('S'), |_| CardinalDirection::South),
+	))(input)?;
+	Ok((i, direction))
+}
+
+/// Parses horizontal directions (East or West). Callers may want to
+/// parse horizontal separate from vertical because you can't parse
+/// two horizontal directions, i.e. E 0.00° W 0.00° is invalid.
+fn horizontal_dir(input: &str) -> IResult<&str, CardinalDirection> {
+	let (i, direction) = alt((
+		map(char('E'), |_| CardinalDirection::East),
+		map(char('W'), |_| CardinalDirection::West),
+	))(input)?;
+	Ok((i, direction))
+}
+
+/// Parses a floating point number, optionally ending in a ° symbol.
+fn degree(input: &str) -> IResult<&str, f64> {
+	let (i, degrees) = double(input)?;
+	let (i, _) = char('°')(i)?;
+	Ok((i, degrees))
+}
+
+/// Parses a degree cardinal (e.g. 40.6892°N) or a cardinal degree (e.g. N40.6892°).
+/// The direction can either be before or after the degree. Callers should provide the
+/// [`CardinalDirection`] parser. This is required because if the caller has already
+/// parsed the vertical direction, they shouldn't be able to parse the horizontal
+/// direction.
+fn cardinal_degree<'i: 't, 't>(
+	direction: impl Fn() -> Box<dyn Parser<&'i str, CardinalDirection, Error<&'i str>>>,
+) -> impl Parser<&'i str, CardinalDegree, Error<&'i str>> {
+	move |input: &'i str| {
+		alt((
+			map(tuple((direction(), mightbespace, degree)), |(dir, _, deg)| {
+				CardinalDegree(dir, deg)
+			}),
+			map(tuple((degree, direction())), |(deg, dir)| CardinalDegree(dir, deg)),
+		))(input)
+	}
+}
```

Comment:
I fought with figuring out how to pass in the `direction` parser to this parser for quite a while. This is what I came up with but would love if there were a more elegant solution.

Reply:
A few ideas:
1. [Function pointer](https://doc.rust-lang.org/reference/types/function-pointer.html) (of the parser) instead of `impl Fn() -> Box<dyn ...>` (a function that returns the parser)
2. I don't see why this has to return a `Parser`; it could probably *be* a parser, like `decimal_degree`.

Reply:
@finnbear updated. Are those changes what you were meaning? I'm still building my Rust knowledge...